changelogs/current.yaml

```diff
@@ -66,6 +66,14 @@ bug_fixes:
     Fixed the bug that updating :ref:`scope_key_builder
     <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.ScopedRoutes.scope_key_builder>`
     of SRDS config doesn't work and multiple HCM share the same ``scope_key_builder``.
+- area: http
+  change: |
+    The :ref:`is_optional <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpFilter.is_optional>` field of
+    HTTP filter can only be used for configuration loading of
+    :ref:`HTTP filter <envoy_v3_api_msg_extensions.filters.network.http_connection_manager.v3.HttpFilter>` and will be ignored for
+    loading of route or virtual host level filter config.
```

Comment:
sorry I'm confused what motivated this change.
I would assume (though API shepherds can correct me) that the optional tag was so one could use the same config for Envoy and gRPC xDS endpoints, where gRPC might not have certain filters implemented or vice versa.  don't we then want is_optional to err on the side of not validating rater than validating?
either way I'm going to ask Adi do primary review as I think it's a question of API and implementation

Reply:
This PR is definitely the right thing from an xDS API perspective.

The purpose of `is_optional` is to allow control planes to configure new filters that may not yet be supported by all existing clients, regardless of whether those clients are Envoy or gRPC or anything else.  The `is_optional` field exists both in the HCM config for the top-level filter configs and also in the `typed_per_filter_config` overrides that exist at the VH, route, and cluster-weight levels.

Envoy's current behavior is that it uses the `is_optional` field at the HCM level when validating the RDS resource, rather than looking at the `is_optional` field in the RDS resource itself.  That behavior is totally broken with regard to resource cacheability, as per discussion in #26097.  This PR fixes Envoy to stop using the `is_optional` field at the HCM level to mark a filter as optional in `typed_per_filter_config` in the RDS resource.

Envoy currently does not use the `is_optional` field in `typed_per_filter_config` at all.  That problem is being fixed in #27210.

My only concern here is that we probably shouldn't leave users without any way to mark a filter as optional in `typed_per_filter_config`.  I think that means that we either need to merge #27210 before this PR, or we need to combine the two into a single PR.

Reply:
I think it would be more easy to review (and to cherry for third party folks) to keep these change in two different PR.
So, if no special requirement, I prefer to land #27210 first.

Reply:
Maybe I can split the #27210 to two pr, one for `is_optional` flag, one for `disabled` flag. The former one is simple and harmless. The latter one will bring some performance overhead and may need more consideration.
I will do it.

Reply:
Good idea!  The `disabled` part of #27210 is not related to this PR, only the `is_optional` bit is, so splitting them up makes sense.

Once you've done that, I think there are the same two choices here: You can either include the `is_optional` field in this PR, or you can put that in a separate PR that gets landed before this one does.  Either one is fine with me.

Thanks!

Reply:
I will put the splited one into this PR. Then all things about the optional extension will be aggregated to this PR.